File: Generator/Archives.cs
/*
 * Author: Pollazzon Stefano
 * Project: ZDoom Navmesh builder
 * This module contains the code for managing 
 * both WAD and PK3 file formats, and supporting
 * Doom, Hexen and UDMF map formats
 */
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Text;

namespace NavmeshBuilder;

internal class TArchiveEntry : Object
{
    internal int EntryOffset { get; set; }
    internal int EntrySize { get; set; }
    internal string EntryName { get; set; }
    internal TArchiveEntry(string EntryName, int EntryOffset, int EntrySize)
    {
        this.EntryName = EntryName;
        this.EntryOffset = EntryOffset;
        this.EntrySize = EntrySize;
    }
}
/// <summary>
/// Class <c>TArchive</c> represents an archive, that can contain multiple maps.
/// Stores all maps in a sorted list; read files from .TXT (UDMF only), .WAD or .PK3 / .ZIP.
/// </summary>
public class TArchive : Object
{
    private List<TMapDefinition> FMapDefinitions;
    public string? FileName { get; private set; }
    public string? FileSpec { get; private set; }
    public long FileSize { get; private set; }
    public DateTime LastUpdate { get; private set; }
    /// <summary>
    /// Property <c>MapDefinitions</c> gives access to the maps in the archive.
    /// </summary>
    public List<TMapDefinition> MapDefinitions { get => FMapDefinitions; }
    private static readonly string[] MapLumpNamesDefs = new string[] {
        "BEHAVIOR",
        "BLOCKMAP",
        "DIALOGUE",
        "ENDMAP",
        "LINEDEFS",
        "NODES",
        "REJECT",
        "SCRIPTS",
        "SECTORS",
        "SEGS",
        "SIDEDEFS",
        "SSECTORS",
        "TEXTMAP",
        "THINGS",
        "VERTEXES",
        "ZNODES"
    };
    public TArchive()
    {
        FMapDefinitions = new List<TMapDefinition>();
    }
    /// <summary>
    /// Function <c>AddMap</c> adds a map.
    /// </summary>
    /// <param name="MapDefinition"><c>MapDefinition</c> is the map to be added.</param>
    public void AddMap(TMapDefinition MapDefinition)
    {
        FMapDefinitions.Add(MapDefinition);
    }
    /// <summary>
    /// Function <c>SortMap</c> sorts the maps in alphabetical order on map names.
    /// </summary>
    /// <param name="MapDefinition"><c>MapDefinition</c> is the map to be added.</param>
    public void SortMap()
    {
        FMapDefinitions.Sort(delegate (TMapDefinition A, TMapDefinition B) { return String.Compare(A.MapName, B.MapName); });
    }
    /// <summary>
    /// Function <c>LoadFromText</c> reads a .TXT file, parsing the UDMF namespace and map entities
    /// </summary>
    /// <param name="FileSpec"><c>FileSpec</c> is the file to be read.</param>
    internal bool LoadFromText(string FileSpec)
    {
        if (File.Exists(FileSpec))
            using (FileStream ArchiveFileStream = File.OpenRead(FileSpec))
            {
                this.FileSpec = FileSpec;
                FileName = Path.GetFileName(FileSpec);
                FileSize = ArchiveFileStream.Length;
                LastUpdate = File.GetLastWriteTime(FileSpec);
                using (StreamReader Reader = new StreamReader(ArchiveFileStream))
                {
                    TUdmfParser UdmfParser = new TUdmfParser(Reader.ReadToEnd());
                    TMapDefinition MapDefinition = new TMapDefinition(Path.GetFileNameWithoutExtension(FileSpec));
                    UdmfParser.ParseMap(MapDefinition);
                    AddMap(MapDefinition);
                }
                return true;
            }
        else
            return false;
    }
    /// <summary>
    /// Function <c>LoadVertexes</c> reads all the VERTEX data in a map.
    /// </summary>
    /// <param name="MapNamespace"><c>MapNamespace</c> is the namespace of the map.</param>
    /// <param name="LumpContent"><c>LumpContent</c> is the binary content of the lump map.</param>
    /// <param name="MapVertex"><c>MapVertex</c> is the list of VERTEX objects of the map, to append the read data.</param>
    internal void LoadVertexes(TMapNamespace MapNamespace, MemoryStream LumpContent, List<TMapVertex> MapVertex)
    {
        MapVertex.Clear();
        BinaryReader LumpBinaryReader = new BinaryReader(LumpContent);
        int Count = Convert.ToInt32(LumpContent.Length / 4);
        for (int I = Count; I > 0; I--)
        {
            TMapVertex Vertex = new TMapVertex(LumpBinaryReader.ReadInt16(), LumpBinaryReader.ReadInt16());
            MapVertex.Add(Vertex);
        }
    }
    /// <summary>
    /// Function <c>LoadLinedefs</c> reads all the LINEDEF data in a map.
    /// </summary>
    /// <param name="MapNamespace"><c>MapNamespace</c> is the namespace of the map.</param>
    /// <param name="LumpContent"><c>LumpContent</c> is the binary content of the lump map.</param>
    /// <param name="MapLinedef"><c>MapLinedef</c> is the list of LINEDEF objects of the map, to append the read data.</param>
    internal void LoadLinedefs(TMapNamespace MapNamespace, MemoryStream LumpContent, List<TMapLinedef> MapLinedef)
    {
        MapLinedef.Clear();
        BinaryReader LumpBinaryReader = new BinaryReader(LumpContent);
        if (MapNamespace == TMapNamespace.MapNamespaceDoom)
        {
            int Count = Convert.ToInt32(LumpContent.Length / 14);
            for (int I = Count; I > 0; I--)
            {
                short V1 = LumpBinaryReader.ReadInt16();
                short V2 = LumpBinaryReader.ReadInt16();
                short Flags = LumpBinaryReader.ReadInt16();
                short Special = LumpBinaryReader.ReadInt16();
                short Arg0 = LumpBinaryReader.ReadInt16();
                short SideFront = LumpBinaryReader.ReadInt16();
                short SideBack = LumpBinaryReader.ReadInt16();
                TMapLinedef Linedef = new TMapLinedef(V1, V2, 0, Flags, SideFront, SideBack, Special, Arg0, 0, 0, 0, 0, false, false, false);
                MapLinedef.Add(Linedef);
            }
        }
        else
        {
            int Count = Convert.ToInt32(LumpContent.Length / 16);
            for (int I = Count; I > 0; I--)
            {
                short V1 = LumpBinaryReader.ReadInt16();
                short V2 = LumpBinaryReader.ReadInt16();
                short Flags = LumpBinaryReader.ReadInt16();
                short Special = Convert.ToInt16(LumpBinaryReader.ReadByte());
                short Arg0 = Convert.ToInt16(LumpBinaryReader.ReadByte());
                short Arg1 = Convert.ToInt16(LumpBinaryReader.ReadByte());
                short Arg2 = Convert.ToInt16(LumpBinaryReader.ReadByte());
                short Arg3 = Convert.ToInt16(LumpBinaryReader.ReadByte());
                short Arg4 = Convert.ToInt16(LumpBinaryReader.ReadByte());
                short SideFront = LumpBinaryReader.ReadInt16();
                short SideBack = LumpBinaryReader.ReadInt16();
                TMapLinedef Linedef = new TMapLinedef(V1, V2, 0, Flags, SideFront, SideBack, Special, Arg0, Arg1, Arg2, Arg3, Arg4, false, false, false);
                MapLinedef.Add(Linedef);
            }
        }
    }
    /// <summary>
    /// Function <c>LoadSidedefs</c> reads all the SIDEDEF data in a map.
    /// </summary>
    /// <param name="MapNamespace"><c>MapNamespace</c> is the namespace of the map.</param>
    /// <param name="LumpContent"><c>LumpContent</c> is the binary content of the lump map.</param>
    /// <param name="MapSidedef"><c>MapSidedef</c> is the list of SIDEDEF objects of the map, to append the read data.</param>
    internal void LoadSidedefs(TMapNamespace MapNamespace, MemoryStream LumpContent, List<TMapSidedef> MapSidedef)
    {
        MapSidedef.Clear();
        BinaryReader LumpBinaryReader = new BinaryReader(LumpContent);
        int Count = Convert.ToInt32(LumpContent.Length / 30);
        for (int I = Count; I > 0; I--)
        {
            LumpBinaryReader.BaseStream.Seek(28, SeekOrigin.Current);
            TMapSidedef Sidedef = new TMapSidedef(LumpBinaryReader.ReadInt16());
            MapSidedef.Add(Sidedef);
        }
    }
    /// <summary>
    /// Function <c>LoadSectors</c> reads all the SECTOR data in a map.
    /// </summary>
    /// <param name="MapNamespace"><c>MapNamespace</c> is the namespace of the map.</param>
    /// <param name="LumpContent"><c>LumpContent</c> is the binary content of the lump map.</param>
    /// <param name="MapSector"><c>MapSector</c> is the list of SECTOR objects of the map, to append the read data.</param>
    internal void LoadSectors(TMapNamespace MapNamespace, MemoryStream LumpContent, List<TMapSector> MapSector)
    {
        MapSector.Clear();
        BinaryReader LumpBinaryReader = new BinaryReader(LumpContent);
        int Count = Convert.ToInt32(LumpContent.Length / 26);
        for (int I = Count; I > 0; I--)
        {
            short HeightFloor = LumpBinaryReader.ReadInt16();
            short HeightCeiling = LumpBinaryReader.ReadInt16();
            LumpBinaryReader.BaseStream.Seek(18, SeekOrigin.Current);
            short Special = LumpBinaryReader.ReadInt16();
            short ID = LumpBinaryReader.ReadInt16();
            TMapSector Sector = new TMapSector(HeightFloor, HeightCeiling, ID, Special, false);
            MapSector.Add(Sector);
        }
    }
    /// <summary>
    /// Function <c>LoadThings</c> reads all the THING data in a map.
    /// </summary>
    /// <param name="MapNamespace"><c>MapNamespace</c> is the namespace of the map.</param>
    /// <param name="LumpContent"><c>LumpContent</c> is the binary content of the lump map.</param>
    /// <param name="MapThing"><c>MapThing</c> is the list of THING objects of the map, to append the read data.</param>
    internal void LoadThings(TMapNamespace MapNamespace, MemoryStream LumpContent, List<TMapThing> MapThing)
    {
        MapThing.Clear();
        BinaryReader LumpBinaryReader = new BinaryReader(LumpContent);
        if (MapNamespace == TMapNamespace.MapNamespaceDoom)
        {
            int Count = Convert.ToInt32(LumpContent.Length / 10);
            for (int I = Count; I > 0; I--)
            {
                short X = LumpBinaryReader.ReadInt16();
                short Y = LumpBinaryReader.ReadInt16();
                LumpBinaryReader.BaseStream.Seek(2, SeekOrigin.Current);
                short ThingType = LumpBinaryReader.ReadInt16();
                LumpBinaryReader.BaseStream.Seek(2, SeekOrigin.Current);
                TMapThing Thing = new TMapThing(X, Y, 0, ThingType, 0, 0, 0, 0, 0, 0, 0);
                MapThing.Add(Thing);
            }
        }
        else
        {
            int Count = Convert.ToInt32(LumpContent.Length / 20);
            for (int I = Count; I > 0; I--)
            {
                short ID = LumpBinaryReader.ReadInt16();
                short X = LumpBinaryReader.ReadInt16();
                short Y = LumpBinaryReader.ReadInt16();
                LumpBinaryReader.BaseStream.Seek(4, SeekOrigin.Current);
                short ThingType = LumpBinaryReader.ReadInt16();
                LumpBinaryReader.BaseStream.Seek(2, SeekOrigin.Current);
                short Special = Convert.ToInt16(LumpBinaryReader.ReadByte());
                short Arg0 = Convert.ToInt16(LumpBinaryReader.ReadByte());
                short Arg1 = Convert.ToInt16(LumpBinaryReader.ReadByte());
                short Arg2 = Convert.ToInt16(LumpBinaryReader.ReadByte());
                short Arg3 = Convert.ToInt16(LumpBinaryReader.ReadByte());
                short Arg4 = Convert.ToInt16(LumpBinaryReader.ReadByte());
                TMapThing Thing = new TMapThing(X, Y, 0, ThingType, ID, Special, Arg0, Arg1, Arg2, Arg3, Arg4);
                MapThing.Add(Thing);
            }
        }
    }
    /// <summary>
    /// Function <c>LoadFromWad</c> loads all the maps contained in a .WAD file.
    /// </summary>
    /// <param name="ArchiveFileStream"><c>ArchiveFileStream</c> is a stream, that gives the binary content of the WAD archive.</param>
    internal bool LoadFromWad(Stream ArchiveFileStream)
    {
        List<string> MapLumpNames = new List<string>(MapLumpNamesDefs);
        using (BinaryReader Reader = new BinaryReader(ArchiveFileStream))
        {
            byte[] HeaderSignature = Reader.ReadBytes(4);
            string HeaderSignatureText = Encoding.ASCII.GetString(HeaderSignature);
            if ((string.Compare(HeaderSignatureText, "IWAD", false) != 0) && (string.Compare(HeaderSignatureText, "PWAD", false) != 0))
                return false;
            int HeaderNumberOfEntries = Reader.ReadInt32();
            int HeaderDirectoryOffset = Reader.ReadInt32();
            Reader.BaseStream.Seek(HeaderDirectoryOffset, SeekOrigin.Begin);
            TArchiveEntry[] ArchiveEntries = new TArchiveEntry[HeaderNumberOfEntries];
            for (int I = 0; I < HeaderNumberOfEntries; I++)
            {
                int EntryOffset = Reader.ReadInt32();
                int EntrySize = Reader.ReadInt32();
                byte[] EntryName = Reader.ReadBytes(8);
                string EntryNameText = Encoding.ASCII.GetString(EntryName).Replace("\0", "");
                ArchiveEntries[I] = new TArchiveEntry(EntryNameText, EntryOffset, EntrySize);
            }
            bool MapFound = false;
            TMapNamespace Namespace = TMapNamespace.MapNamespaceDoom;
            string? TextMap = null;
            List<TMapVertex> MapVertex = new List<TMapVertex>();
            List<TMapLinedef> MapLinedef = new List<TMapLinedef>();
            List<TMapSidedef> MapSidedef = new List<TMapSidedef>();
            List<TMapSector> MapSector = new List<TMapSector>();
            List<TMapThing> MapThing = new List<TMapThing>();
            int EntryNumber = HeaderNumberOfEntries - 1;
            while (EntryNumber >= 0)
            {
                if (MapLumpNames.Contains(ArchiveEntries[EntryNumber].EntryName))
                {
                    MapFound = true;
                    if (string.Compare(ArchiveEntries[EntryNumber].EntryName, "BEHAVIOR", false) == 0)
                        Namespace = TMapNamespace.MapNamespaceHexen;
                    if (string.Compare(ArchiveEntries[EntryNumber].EntryName, "TEXTMAP", false) == 0)
                    {
                        long SavedPosition = Reader.BaseStream.Position;
                        Reader.BaseStream.Position = ArchiveEntries[EntryNumber].EntryOffset;
                        byte[] LumpContent = Reader.ReadBytes(ArchiveEntries[EntryNumber].EntrySize);
                        TextMap = Encoding.UTF8.GetString(LumpContent);
                        Reader.BaseStream.Position = SavedPosition;
                    }
                    if (string.Compare(ArchiveEntries[EntryNumber].EntryName, "VERTEXES", false) == 0)
                    {
                        long SavedPosition = Reader.BaseStream.Position;
                        Reader.BaseStream.Position = ArchiveEntries[EntryNumber].EntryOffset;
                        byte[] LumpContent = Reader.ReadBytes(ArchiveEntries[EntryNumber].EntrySize);
                        MemoryStream MemoryLump = new MemoryStream(LumpContent);
                        LoadVertexes(Namespace, MemoryLump, MapVertex);
                        Reader.BaseStream.Position = SavedPosition;
                    }
                    if (string.Compare(ArchiveEntries[EntryNumber].EntryName, "LINEDEFS", false) == 0)
                    {
                        long SavedPosition = Reader.BaseStream.Position;
                        Reader.BaseStream.Position = ArchiveEntries[EntryNumber].EntryOffset;
                        byte[] LumpContent = Reader.ReadBytes(ArchiveEntries[EntryNumber].EntrySize);
                        MemoryStream MemoryLump = new MemoryStream(LumpContent);
                        LoadLinedefs(Namespace, MemoryLump, MapLinedef);
                        Reader.BaseStream.Position = SavedPosition;
                    }
                    if (string.Compare(ArchiveEntries[EntryNumber].EntryName, "SIDEDEFS", false) == 0)
                    {
                        long SavedPosition = Reader.BaseStream.Position;
                        Reader.BaseStream.Position = ArchiveEntries[EntryNumber].EntryOffset;
                        byte[] LumpContent = Reader.ReadBytes(ArchiveEntries[EntryNumber].EntrySize);
                        MemoryStream MemoryLump = new MemoryStream(LumpContent);
                        LoadSidedefs(Namespace, MemoryLump, MapSidedef);
                        Reader.BaseStream.Position = SavedPosition;
                    }
                    if (string.Compare(ArchiveEntries[EntryNumber].EntryName, "SECTORS", false) == 0)
                    {
                        long SavedPosition = Reader.BaseStream.Position;
                        Reader.BaseStream.Position = ArchiveEntries[EntryNumber].EntryOffset;
                        byte[] LumpContent = Reader.ReadBytes(ArchiveEntries[EntryNumber].EntrySize);
                        MemoryStream MemoryLump = new MemoryStream(LumpContent);
                        LoadSectors(Namespace, MemoryLump, MapSector);
                        Reader.BaseStream.Position = SavedPosition;
                    }
                    if (string.Compare(ArchiveEntries[EntryNumber].EntryName, "THINGS", false) == 0)
                    {
                        long SavedPosition = Reader.BaseStream.Position;
                        Reader.BaseStream.Position = ArchiveEntries[EntryNumber].EntryOffset;
                        byte[] LumpContent = Reader.ReadBytes(ArchiveEntries[EntryNumber].EntrySize);
                        MemoryStream MemoryLump = new MemoryStream(LumpContent);
                        LoadThings(Namespace, MemoryLump, MapThing);
                        Reader.BaseStream.Position = SavedPosition;
                    }
                }
                else
                {
                    if (MapFound)
                    {
                        string MapName = ArchiveEntries[EntryNumber].EntryName;
                        TMapDefinition MapDefinition = new TMapDefinition(MapName);
                        if (!string.IsNullOrEmpty(TextMap))
                        {
                            TUdmfParser UdmfParser = new TUdmfParser(TextMap);
                            UdmfParser.ParseMap(MapDefinition);
                            MapDefinition.MapNamespace = UdmfParser.MapNamespace;
                            AddMap(MapDefinition);
                            TextMap = null;
                        }
                        else
                        {
                            MapVertex.ForEach(MapDefinition.AddMapVertex);
                            MapLinedef.ForEach(MapDefinition.AddMapLinedef);
                            MapSidedef.ForEach(MapDefinition.AddMapSidedef);
                            MapSector.ForEach(MapDefinition.AddMapSector);
                            MapThing.ForEach(MapDefinition.AddMapThing);
                            AddMap(MapDefinition);
                        }
                        MapFound = false;
                        Namespace = TMapNamespace.MapNamespaceDoom;
                    }
                }
                EntryNumber--;
            }
        }
        SortMap();
        return true;
    }
    /// <summary>
    /// Function <c>LoadFromWad</c> loads all the maps contained in a .WAD file.
    /// </summary>
    /// <param name="FileSpec"><c>FileSpec</c> is the file to be read.</param>
    internal bool LoadFromWad(string FileSpec)
    {
        if (File.Exists(FileSpec))
            using (FileStream ArchiveFileStream = File.OpenRead(FileSpec))
            {
                this.FileSpec = FileSpec;
                FileName = Path.GetFileName(FileSpec);
                FileSize = ArchiveFileStream.Length;
                LastUpdate = File.GetLastWriteTime(FileSpec);
                return LoadFromWad(ArchiveFileStream);
            }
        else
            return false;
    }
    /// <summary>
    /// Function <c>LoadFromWad</c> loads all the maps contained in a .PK3 / .ZIP file.
    /// All maps must be stored in the /maps/ folder and be in .WAD format.
    /// </summary>
    /// <param name="FileSpec"><c>FileSpec</c> is the file to be read.</param>
    internal bool LoadFromZip(string FileSpec)
    {
        if (File.Exists(FileSpec))
            using (FileStream ArchiveFileStream = File.OpenRead(FileSpec))
            {
                this.FileSpec = FileSpec;
                FileName = Path.GetFileName(FileSpec);
                FileSize = ArchiveFileStream.Length;
                LastUpdate = File.GetLastWriteTime(FileSpec);
                using (ZipArchive Archive = new ZipArchive(ArchiveFileStream, ZipArchiveMode.Read))
                {
                    foreach (ZipArchiveEntry ArchiveEntry in Archive.Entries)
                        if ((ArchiveEntry.FullName.StartsWith("MAPS/", StringComparison.OrdinalIgnoreCase) && ArchiveEntry.FullName.EndsWith(".WAD", StringComparison.OrdinalIgnoreCase)))
                        {
                            using (Stream MapStream = ArchiveEntry.Open())
                            {
                                MemoryStream SeekableMapStream = new MemoryStream();
                                MapStream.CopyTo(SeekableMapStream);
                                SeekableMapStream.Position = 0;
                                LoadFromWad(SeekableMapStream);
                            }
                        }
                }
                return true;
            }
        else
            return false;
    }
    /// <summary>
    /// Function <c>Load</c> loads all the maps contained in a .PK3 / .ZIP file.
    /// All maps must be stored in the /maps/ folder and be in .WAD format.
    /// </summary>
    /// <param name="FileSpec"><c>FileSpec</c> is the file to be read.</param>
    public bool Load(string FileSpec)
    {
        if (File.Exists(FileSpec))
        {
            string FileExtension = Path.GetExtension(FileSpec);
            if (string.Compare(FileExtension, ".TXT", true) == 0)
                return LoadFromText(FileSpec);
            if (string.Compare(FileExtension, ".WAD", true) == 0)
                return LoadFromWad(FileSpec);
            if ((string.Compare(FileExtension, ".PK3", true) == 0) || (string.Compare(FileExtension, ".ZIP", true) == 0))
                return LoadFromZip(FileSpec);
            return false;
        }
        else
            return false;
    }
}

File: Generator/Python/Archives.py
"""
- Author: Pollazzon Stefano (ported)
- Project: ZDoom Navmesh builder
- This module contains the code for managing
- both WAD and PK3 file formats, and supporting
- Doom, Hexen and UDMF map formats
"""

import io
import os
import struct
import zipfile

from Structures import (
    TMapVertex,
    TMapLinedef,
    TMapSidedef,
    TMapSector,
    TMapThing,
    TMapDefinition,
    TMapNamespace,
)
from Parsers import TUdmfParser


class TArchiveEntry:
    def __init__(self, EntryName, EntryOffset, EntrySize):
        self.EntryName = EntryName
        self.EntryOffset = int(EntryOffset)
        self.EntrySize = int(EntrySize)


class TArchive:
    MapLumpNamesDefs = [
        "BEHAVIOR",
        "BLOCKMAP",
        "DIALOGUE",
        "ENDMAP",
        "LINEDEFS",
        "NODES",
        "REJECT",
        "SCRIPTS",
        "SECTORS",
        "SEGS",
        "SIDEDEFS",
        "SSECTORS",
        "TEXTMAP",
        "THINGS",
        "VERTEXES",
        "ZNODES",
    ]

    def __init__(self):
        self.MapDefinitions = []
        self.FileName = ""
        self.FileSpec = ""
        self.FileSize = 0
        self.LastUpdate = 0.0

    # --------------------------------------------------
    # Helpers
    # --------------------------------------------------

    def AddMap(self, MapDefinition):
        self.MapDefinitions.append(MapDefinition)

    def SortMap(self):
        self.MapDefinitions.sort(key=lambda m: m.MapName)

    # --------------------------------------------------
    # Text maps (UDMF only)
    # --------------------------------------------------

    def LoadFromText(self, FileSpec):
        if not os.path.exists(FileSpec):
            return False

        self.FileSpec = FileSpec
        self.FileName = os.path.basename(FileSpec)
        self.FileSize = os.path.getsize(FileSpec)
        self.LastUpdate = os.path.getmtime(FileSpec)

        with open(FileSpec, "r", encoding="utf-8") as reader:
            text_map = reader.read()

        udmf_parser = TUdmfParser(text_map)
        map_definition = TMapDefinition(os.path.splitext(os.path.basename(FileSpec))[0])
        udmf_parser.ParseMap(map_definition)
        self.AddMap(map_definition)
        return True

    # --------------------------------------------------
    # Binary lump readers
    # --------------------------------------------------

    def LoadVertexes(self, MapNamespaceValue, LumpContent, MapVertex):
        MapVertex.clear()
        count = len(LumpContent) // 4
        offset = 0
        for _ in range(count):
            x, y = struct.unpack_from("<hh", LumpContent, offset)
            offset += 4
            MapVertex.append(TMapVertex(x, y))

    def LoadLinedefs(self, MapNamespaceValue, LumpContent, MapLinedef):
        MapLinedef.clear()
        offset = 0

        if MapNamespaceValue == TMapNamespace.MapNamespaceDoom:
            count = len(LumpContent) // 14
            for _ in range(count):
                v1, v2, flags, special, arg0, side_front, side_back = struct.unpack_from(
                    "<hhhhhhh", LumpContent, offset
                )
                offset += 14

                linedef = TMapLinedef(
                    V1=v1,
                    V2=v2,
                    ID=0,
                    Blocking=((flags & 0x0003) != 0),
                    TwoSided=((flags & 0x0004) != 0),
                    SideFront=side_front,
                    SideBack=side_back,
                    Special=special,
                    Arg0=arg0,
                    Arg1=0,
                    Arg2=0,
                    Arg3=0,
                    Arg4=0,
                    MonsterUse=False,
                    RepeatableSpecial=False,
                    Ignored=False,
                )
                MapLinedef.append(linedef)
        else:
            count = len(LumpContent) // 16
            for _ in range(count):
                v1, v2, flags = struct.unpack_from("<hhh", LumpContent, offset)
                offset += 6

                special, arg0, arg1, arg2, arg3, arg4 = struct.unpack_from("<BBBBBB", LumpContent, offset)
                offset += 6

                side_front, side_back = struct.unpack_from("<hh", LumpContent, offset)
                offset += 4

                linedef = TMapLinedef(
                    V1=v1,
                    V2=v2,
                    ID=0,
                    Blocking=((flags & 0x0003) != 0),
                    TwoSided=((flags & 0x0004) != 0),
                    SideFront=side_front,
                    SideBack=side_back,
                    Special=special,
                    Arg0=arg0,
                    Arg1=arg1,
                    Arg2=arg2,
                    Arg3=arg3,
                    Arg4=arg4,
                    MonsterUse=False,
                    RepeatableSpecial=False,
                    Ignored=False,
                )
                MapLinedef.append(linedef)

    def LoadSidedefs(self, MapNamespaceValue, LumpContent, MapSidedef):
        MapSidedef.clear()
        count = len(LumpContent) // 30
        offset = 0

        for _ in range(count):
            # skip texture / offset data, sector index is last 2 bytes
            sector = struct.unpack_from("<h", LumpContent, offset + 28)[0]
            offset += 30
            MapSidedef.append(TMapSidedef(sector))

    def LoadSectors(self, MapNamespaceValue, LumpContent, MapSector):
        MapSector.clear()
        count = len(LumpContent) // 26
        offset = 0

        for _ in range(count):
            height_floor, height_ceiling = struct.unpack_from("<hh", LumpContent, offset)
            offset += 4

            # skip 18 bytes of floor/ceiling textures + light + type-dependent data before special/tag
            offset += 18

            special, sector_id = struct.unpack_from("<hh", LumpContent, offset)
            offset += 4

            MapSector.append(TMapSector(height_floor, height_ceiling, sector_id, special, False))

    def LoadThings(self, MapNamespaceValue, LumpContent, MapThing):
        MapThing.clear()
        offset = 0

        if MapNamespaceValue == TMapNamespace.MapNamespaceDoom:
            count = len(LumpContent) // 10
            for _ in range(count):
                x, y = struct.unpack_from("<hh", LumpContent, offset)
                offset += 4

                offset += 2  # angle
                thing_type = struct.unpack_from("<h", LumpContent, offset)[0]
                offset += 2
                offset += 2  # flags

                MapThing.append(TMapThing(x, y, 0, thing_type, 0, 0, 0, 0, 0, 0, 0))
        else:
            count = len(LumpContent) // 20
            for _ in range(count):
                thing_id, x, y = struct.unpack_from("<hhh", LumpContent, offset)
                offset += 6

                offset += 4  # height + angle
                thing_type = struct.unpack_from("<h", LumpContent, offset)[0]
                offset += 2

                offset += 2  # flags
                special, arg0, arg1, arg2, arg3, arg4 = struct.unpack_from("<BBBBBB", LumpContent, offset)
                offset += 6

                MapThing.append(
                    TMapThing(x, y, 0, thing_type, thing_id, special, arg0, arg1, arg2, arg3, arg4)
                )

    # --------------------------------------------------
    # WAD loading
    # --------------------------------------------------

    def LoadFromWadStream(self, ArchiveFileStream):
        map_lump_names = list(self.MapLumpNamesDefs)

        reader = ArchiveFileStream
        reader.seek(0)

        header_signature = reader.read(4)
        header_signature_text = header_signature.decode("ascii", errors="ignore")

        if header_signature_text not in ("IWAD", "PWAD"):
            return False

        header_number_of_entries = struct.unpack("<i", reader.read(4))[0]
        header_directory_offset = struct.unpack("<i", reader.read(4))[0]

        reader.seek(header_directory_offset)
        archive_entries = []

        for _ in range(header_number_of_entries):
            entry_offset, entry_size = struct.unpack("<ii", reader.read(8))
            entry_name_raw = reader.read(8)
            entry_name_text = entry_name_raw.decode("ascii", errors="ignore").replace("\x00", "")
            archive_entries.append(TArchiveEntry(entry_name_text, entry_offset, entry_size))

        map_found = False
        namespace = TMapNamespace.MapNamespaceDoom
        text_map = None

        map_vertex = []
        map_linedef = []
        map_sidedef = []
        map_sector = []
        map_thing = []

        entry_number = header_number_of_entries - 1
        while entry_number >= 0:
            entry = archive_entries[entry_number]

            if entry.EntryName in map_lump_names:
                map_found = True

                if entry.EntryName == "BEHAVIOR":
                    namespace = TMapNamespace.MapNamespaceHexen

                if entry.EntryName == "TEXTMAP":
                    saved_position = reader.tell()
                    reader.seek(entry.EntryOffset)
                    lump_content = reader.read(entry.EntrySize)
                    text_map = lump_content.decode("utf-8", errors="ignore")
                    reader.seek(saved_position)

                if entry.EntryName == "VERTEXES":
                    saved_position = reader.tell()
                    reader.seek(entry.EntryOffset)
                    lump_content = reader.read(entry.EntrySize)
                    self.LoadVertexes(namespace, lump_content, map_vertex)
                    reader.seek(saved_position)

                if entry.EntryName == "LINEDEFS":
                    saved_position = reader.tell()
                    reader.seek(entry.EntryOffset)
                    lump_content = reader.read(entry.EntrySize)
                    self.LoadLinedefs(namespace, lump_content, map_linedef)
                    reader.seek(saved_position)

                if entry.EntryName == "SIDEDEFS":
                    saved_position = reader.tell()
                    reader.seek(entry.EntryOffset)
                    lump_content = reader.read(entry.EntrySize)
                    self.LoadSidedefs(namespace, lump_content, map_sidedef)
                    reader.seek(saved_position)

                if entry.EntryName == "SECTORS":
                    saved_position = reader.tell()
                    reader.seek(entry.EntryOffset)
                    lump_content = reader.read(entry.EntrySize)
                    self.LoadSectors(namespace, lump_content, map_sector)
                    reader.seek(saved_position)

                if entry.EntryName == "THINGS":
                    saved_position = reader.tell()
                    reader.seek(entry.EntryOffset)
                    lump_content = reader.read(entry.EntrySize)
                    self.LoadThings(namespace, lump_content, map_thing)
                    reader.seek(saved_position)

            else:
                if map_found:
                    map_name = entry.EntryName
                    map_definition = TMapDefinition(map_name)

                    if text_map:
                        udmf_parser = TUdmfParser(text_map)
                        udmf_parser.ParseMap(map_definition)
                        map_definition.MapNamespace = udmf_parser.MapNamespace
                        self.AddMap(map_definition)
                        text_map = None
                    else:
                        map_definition.MapNamespace = namespace
                        for item in map_vertex:
                            map_definition.AddMapVertex(item)
                        for item in map_linedef:
                            map_definition.AddMapLinedef(item)
                        for item in map_sidedef:
                            map_definition.AddMapSidedef(item)
                        for item in map_sector:
                            map_definition.AddMapSector(item)
                        for item in map_thing:
                            map_definition.AddMapThing(item)
                        self.AddMap(map_definition)

                    map_found = False
                    namespace = TMapNamespace.MapNamespaceDoom
                    map_vertex = []
                    map_linedef = []
                    map_sidedef = []
                    map_sector = []
                    map_thing = []

            entry_number -= 1

        self.SortMap()
        return True

    def LoadFromWad(self, FileSpec):
        if not os.path.exists(FileSpec):
            return False

        self.FileSpec = FileSpec
        self.FileName = os.path.basename(FileSpec)
        self.FileSize = os.path.getsize(FileSpec)
        self.LastUpdate = os.path.getmtime(FileSpec)

        with open(FileSpec, "rb") as archive_file_stream:
            return self.LoadFromWadStream(archive_file_stream)

    # --------------------------------------------------
    # ZIP / PK3 loading
    # --------------------------------------------------

    def LoadFromZip(self, FileSpec):
        if not os.path.exists(FileSpec):
            return False

        self.FileSpec = FileSpec
        self.FileName = os.path.basename(FileSpec)
        self.FileSize = os.path.getsize(FileSpec)
        self.LastUpdate = os.path.getmtime(FileSpec)

        with zipfile.ZipFile(FileSpec, "r") as archive:
            for archive_entry in archive.infolist():
                full_name = archive_entry.filename

                if full_name.upper().startswith("MAPS/") and full_name.upper().endswith(".WAD"):
                    with archive.open(archive_entry, "r") as map_stream:
                        seekable_map_stream = io.BytesIO(map_stream.read())
                        seekable_map_stream.seek(0)
                        self.LoadFromWadStream(seekable_map_stream)

        return True

    # --------------------------------------------------
    # Dispatch
    # --------------------------------------------------

    def Load(self, FileSpec):
        if not os.path.exists(FileSpec):
            return False

        file_extension = os.path.splitext(FileSpec)[1].upper()

        # reset maps on each explicit Load call
        self.MapDefinitions = []

        if file_extension == ".TXT":
            return self.LoadFromText(FileSpec)

        if file_extension == ".WAD":
            return self.LoadFromWad(FileSpec)

        if file_extension in (".PK3", ".ZIP"):
            return self.LoadFromZip(FileSpec)

        return False